README.md
# rotatingnav

[![Build Status](https://secure.travis-ci.org/audreyr/rotatingnav.png?branch=master)](https://travis-ci.org/audreyr/rotatingnav)

A nav menu that rotates through infinite links, allowing for more nav menu links than can normally fit.

![Screenshot of rotatingnav](rotatingnav-screenshot.png)

The above is just a screenshot. For a live demo, see http://bl.ocks.org/audreyr/6151852. In the demo:

* To advance forward: click > or press f
* To go backward: click < or press b

This will be going into the new djangopackages.com and OpenComparison framework soon. Stay tuned.

### Usage

Add this bit of JS:

```
$(function() {
	$(".rotatingnav").rotatingnav({
    panelCount: 8,
    activeCount: 4
	});
});
```

As for the corresponding HTML, see [demo/index.html](https://github.com/audreyr/rotatingnav/blob/master/demo/index.html) for an example of how to define the HTML for a nav menu with tons and tons of links.

### License

MIT. http://audreyr.mit-license.org


HISTORY.md
# History

* v0.1.0 August 5, 2013
	* First release


carousel.js
/* ========================================================================
 * Bootstrap: carousel.js v3.0.0
 * http://twbs.github.com/bootstrap/javascript.html#carousel
 * ========================================================================
 * Copyright 2012 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ======================================================================== */


+function ($) { "use strict";

  // CAROUSEL CLASS DEFINITION
  // =========================

  var Carousel = function (element, options) {
    this.$element    = $(element)
    this.options     = options
    this.paused      =
    this.sliding     =
    this.interval    =
    this.$active     =
    this.$items      = null

    this.options.pause == 'hover' && this.$element
      .on('mouseenter', $.proxy(this.pause, this))
      .on('mouseleave', $.proxy(this.cycle, this))
  }

  Carousel.DEFAULTS = {
    interval: 5000
  , pause: 'hover'
  }

  Carousel.prototype.cycle =  function (e) {
    e || (this.paused = false)

    this.interval && clearInterval(this.interval)

    this.options.interval
      && !this.paused
      && (this.interval = setInterval($.proxy(this.next, this), this.options.interval))

    return this
  }

  Carousel.prototype.getActiveIndex = function () {
    this.$active = this.$element.find('.item.active')
    this.$items  = this.$active.parent().children()

    return this.$items.index(this.$active)
  }

  Carousel.prototype.to = function (pos) {
    var that        = this
    var activeIndex = this.getActiveIndex()

    if (pos > (this.$items.length - 1) || pos < 0) return

    if (this.sliding)       return this.$element.one('slid', function () { that.to(pos) })
    if (activeIndex == pos) return this.pause().cycle()

    return this.slide(pos > activeIndex ? 'next' : 'prev', $(this.$items[pos]))
  }

  Carousel.prototype.pause = function (e) {
    e || (this.paused = true)

    if (this.$element.find('.next, .prev').length && $.support.transition.end) {
      this.$element.trigger($.support.transition.end)
      this.cycle(true)
    }

    this.interval = clearInterval(this.interval)

    return this
  }

  Carousel.prototype.next = function () {
    if (this.sliding) return
    return this.slide('next')
  }

  Carousel.prototype.prev = function () {
    if (this.sliding) return
    return this.slide('prev')
  }

  Carousel.prototype.slide = function (type, next) {
    var $active   = this.$element.find('.item.active')
    var $active2   = this.$element.find('.item.active2')
    var $next     = next || $active2[type]()
    var isCycling = this.interval
    var direction = type == 'next' ? 'left' : 'right'
    var fallback  = type == 'next' ? 'first' : 'last'
    var that      = this

    this.sliding = true

    isCycling && this.pause()

    $next = $next.length ? $next : this.$element.find('.item')[fallback]()

    var e = $.Event('slide.bs.carousel', { relatedTarget: $next[0], direction: direction })

    if ($next.hasClass('active')) return

    if ($.support.transition && this.$element.hasClass('slide')) {
      this.$element.trigger(e)
      if (e.isDefaultPrevented()) return
      $next.addClass(type)
      $next[0].offsetWidth // force reflow
      $active.addClass(direction)
      $next.addClass(direction)
      $active
        .one($.support.transition.end, function () {
          $active.removeClass(['active', direction].join(' '))
          $active2.removeClass(['active2', direction].join(' ')).addClass('active')
          $next.removeClass([type, direction].join(' ')).addClass('active2')
          that.sliding = false
          setTimeout(function () { that.$element.trigger('slid') }, 0)
        })
        .emulateTransitionEnd(600)
    } else {
      this.$element.trigger(e)
      if (e.isDefaultPrevented()) return
      $active.removeClass('active')
      $active2.addClass('active')
      $active2.removeClass('active2')
      $next.addClass('active2')
      this.sliding = false
      this.$element.trigger('slid')
    }
    
    if ($('.active2').index() < $('.active').index()) {
        $('.active').addClass('pull-left');
    } else {
        $('.active').removeClass('pull-left');
    }

    isCycling && this.cycle()

    return this
  }


  // CAROUSEL PLUGIN DEFINITION
  // ==========================

  var old = $.fn.carousel

  $.fn.carousel = function (option) {
    return this.each(function () {
      var $this   = $(this)
      var data    = $this.data('bs.carousel')
      var options = $.extend({}, Carousel.DEFAULTS, $this.data(), typeof option == 'object' && option)
      var action  = typeof option == 'string' ? option : options.slide

      if (!data) $this.data('bs.carousel', (data = new Carousel(this, options)))
      if (typeof option == 'number') data.to(option)
      else if (action) data[action]()
      else if (options.interval) data.pause().cycle()
    })
  }

  $.fn.carousel.Constructor = Carousel


  // CAROUSEL NO CONFLICT
  // ====================

  $.fn.carousel.noConflict = function () {
    $.fn.carousel = old
    return this
  }


  // CAROUSEL DATA-API
  // =================

  $(document).on('click.bs.carousel.data-api', '[data-slide], [data-slide-to]', function (e) {
    var $this   = $(this), href
    var $target = $($this.attr('data-target') || (href = $this.attr('href')) && href.replace(/.*(?=#[^\s]+$)/, '')) //strip for ie7
    var options = $.extend({}, $target.data(), $this.data())
    var slideIndex = $this.attr('data-slide-to')
    if (slideIndex) options.interval = false

    $target.carousel(options)

    if (slideIndex = $this.attr('data-slide-to')) {
      $target.data('bs.carousel').to(slideIndex)
    }

    e.preventDefault()
  })

  $(window).on('load', function () {
    $('[data-ride="carousel"]').each(function () {
      var $carousel = $(this)
      $carousel.carousel($carousel.data())
    })
  })

}(window.jQuery);


Gruntfile.js
module.exports = function(grunt) {

  grunt.initConfig({

    // Import package manifest
    pkg: grunt.file.readJSON("rotatingnav.jquery.json"),

    // Banner definitions
    meta: {
      banner: "/*\n" +
        " *  <%= pkg.title || pkg.name %> - v<%= pkg.version %>\n" +
        " *  <%= pkg.description %>\n" +
        " *  <%= pkg.homepage %>\n" +
        " *\n" +
        " *  Made by <%= pkg.author.name %>\n" +
        " *  Under <%= pkg.licenses[0].type %> License\n" +
        " */\n"
    },

    // Concat definitions
    concat: {
      dist: {
        src: ["src/jquery.rotatingnav.js"],
        dest: "dist/jquery.rotatingnav.js"
      },
      options: {
        banner: "<%= meta.banner %>"
      }
    },

    // Lint definitions
    jshint: {
      files: ["src/jquery.rotatingnav.js"],
      options: {
        jshintrc: ".jshintrc"
      }
    },
    
    // Copy main js file to dist
    copy: {
      dist: {
        src: "src/jquery.rotatingnav.js",
        dest: "dist/jquery.rotatingnav.js"
      }
    },

    // Minify definitions
    uglify: {
      my_target: {
        src: "dist/jquery.rotatingnav.js",
        dest: "dist/jquery.rotatingnav.min.js"
      },
      options: {
        banner: "<%= meta.banner %>",
        compress: {
          global_defs: {
            "DEBUG": false
          },
          dead_code: true
        }
      }
    },

    // Less compilation
    less: {
      compile: {
        files: {
          "dist/jquery.rotatingnav.css": "src/rotatingnav.less"
        }
      }
    },
    
    // CSS minification
    cssmin: {
      add_banner: {
        options: {
          banner: '/* rotatingnav by audreyr ~ https://github.com/audreyr/rotatingnav */'
        },
        files: {
          'dist/jquery.rotatingnav.min.css': ['dist/jquery.rotatingnav.css']
        }
      }
    },

    // Release to NPM
    release: {
      
    }
    
  });

  grunt.loadNpmTasks("grunt-contrib-concat");
  grunt.loadNpmTasks("grunt-contrib-jshint");
  grunt.loadNpmTasks('grunt-contrib-copy');
  grunt.loadNpmTasks("grunt-contrib-uglify");
  grunt.loadNpmTasks("grunt-contrib-less");
  grunt.loadNpmTasks("grunt-contrib-cssmin");
  grunt.loadNpmTasks('grunt-release');

  grunt.registerTask("default", ["jshint", "concat", "copy", "uglify", "less", "cssmin"]);
  grunt.registerTask("travis", ["jshint"]);

};


rotatingnav.js
var activeHead = 0;
var activeTail = 3;
var panelHead = 0;
var panelTail = 7;

function shiftForward() {
    // Press right arrow
    console.log("Shifting forward");
    activeHead++;
    activeTail++;
  if (activeTail > panelTail) {
    activeTail = panelHead;
  }
  if (activeHead > panelTail) {
    activeHead = panelHead;
  }
  updateActive();
}

function shiftBackward() {
    // Press left arrow
    console.log("Shifting backward");
    activeHead--;
    activeTail--;
  if (activeHead < panelHead) {
    activeHead = panelTail;
  }
  
  if (activeTail < panelHead) {
    activeTail = panelTail;
  }
  updateActive();
}

function isActive(index) {
  if (activeHead < activeTail && 
      activeHead <= index && 
      index <= activeTail) {
    return true;
  } else if (activeHead > activeTail && 
             !(activeTail < index && 
               index < activeHead)) {
    return true;
  } else {
    return false;
  }
}

function updateActive() {
  console.log("activeHead " + activeHead + ", activeTail " + activeTail);
  $('.item').removeClass('pull-left');
  $('.item').removeClass('active');
  $('.item').each(function(index){
    if (isActive(index)) {
      $(this).addClass('active');
      $(this).show();
      if (index > activeTail) {
        $(this).addClass('pull-left');
      }
    } else {
      $(this).hide();
    }
  });
}

$(function() {
  updateActive();
  $('.left.rotatingnav-control').click(shiftForward);
  $('.right.rotatingnav-control').click(shiftForward);
  $('body').keypress(function (event) {
    if (event.which == 102) { // f - forward
      shiftForward();
    } else if (event.which == 98) { // b - backward
      shiftBackward();   
    }
  });
});

src/jquery.rotatingnav.js
/*
 *  jQuery RotatingNav
 *  A nav menu that rotates through infinite links, allowing for more nav menu links than can normally fit.
 *  https://github.com/audreyr/rotatingnav/
 *
 *  Made by Audrey Roy
 *  Under MIT License
 */
if (typeof DEBUG === "undefined") {
  DEBUG = true;
}

;(function ( $, window, document, undefined ) {

  var rotatingnav = "rotatingnav",
      defaults = {
        panelCount: 8,
        activeCount: 4
      };

  function RotatingNav ( element, options ) {
    this.element = element;
    this.settings = $.extend( {}, defaults, options );
    this._defaults = defaults;
    this._name = rotatingnav;
    this.init();
  }

  RotatingNav.prototype = {

    init: function () {
      var that = this;
      that.activeHead = 0;
      that.activeTail = this.settings.activeCount - 1;
      that.panelHead = 0;
      that.panelTail = this.settings.panelCount - 1;

      that.updateActive();
      $(".left.rotatingnav-control").click(function () {
        that.shiftBackward();
      });
      $(".right.rotatingnav-control").click(function () {
        that.shiftForward();
      });
      $("body").keypress(function (event) {
        if (event.which === 102) { // f - forward
          that.shiftForward();
        } else if (event.which === 98) { // b - backward
          that.shiftBackward();
        }
      });
    },

    isActive: function (index) {
      if (this.activeHead < this.activeTail &&
          this.activeHead <= index &&
          index <= this.activeTail) {
        return true;
      } else if (this.activeHead > this.activeTail &&
                 !(this.activeTail < index &&
                   index < this.activeHead)) {
        return true;
      } else {
        return false;
      }
    },

    updateActive: function () {
      var that = this;
      DEBUG && console.log("activeHead " + that.activeHead + ", activeTail " + that.activeTail);
      $(".item").removeClass("pull-left");
      $(".item").removeClass("active");
      $(".item").each(function(index){
        if (that.isActive(index)) {
          $(this).addClass("active");
          $(this).show();
          if (index > that.activeTail) {
            $(this).addClass("pull-left");
          }
        } else {
          $(this).hide();
        }
      });
    },

    shiftForward: function () {
      DEBUG && console.log("Shifting forward");
      this.activeHead++;
      this.activeTail++;
      if (this.activeTail > this.panelTail) {
        this.activeTail = this.panelHead;
      }
      if (this.activeHead > this.panelTail) {
        this.activeHead = this.panelHead;
      }
      this.updateActive();
    },

    shiftBackward: function () {
      DEBUG && console.log("Shifting backward");
      this.activeHead--;
      this.activeTail--;
      if (this.activeHead < this.panelHead) {
        this.activeHead = this.panelTail;
      }

      if (this.activeTail < this.panelHead) {
        this.activeTail = this.panelTail;
      }
      this.updateActive();
    }

  };

  $.fn[ rotatingnav ] = function ( options ) {
    return this.each(function () {
      if ( !$.data( this, "_" + rotatingnav ) ) {
        $.data (this, "plugin_" + rotatingnav, new RotatingNav( this, options ) );
      }
    });
  };

})( jQuery, window, document );


rotating-nav.js
/* ========================================================================
 * Copyright 2013 Audrey Roy
 *
 * Adapted from Bootstrap: carousel.js v3.0.0
 * http://twbs.github.com/bootstrap/javascript.html#carousel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ======================================================================== */


+function ($) { "use strict";

  // CAROUSEL CLASS DEFINITION
  // =========================

  var Carousel = function (element, options) {
    this.$element    = $(element)
    this.options     = options
    this.paused      =
    this.sliding     =
    this.interval    =
    this.$active     =
    this.$items      = null

    this.options.pause == 'hover' && this.$element
      .on('mouseenter', $.proxy(this.pause, this))
      .on('mouseleave', $.proxy(this.cycle, this))
  }

  Carousel.DEFAULTS = {
    interval: 5000
  , pause: 'hover'
  }

  Carousel.prototype.cycle =  function (e) {
    e || (this.paused = false)

    this.interval && clearInterval(this.interval)

    this.options.interval
      && !this.paused
      && (this.interval = setInterval($.proxy(this.next, this), this.options.interval))

    return this
  }

  Carousel.prototype.getActiveIndex = function () {
    this.$active = this.$element.find('.item.active')
    this.$items  = this.$active.parent().children()

    return this.$items.index(this.$active)
  }

  Carousel.prototype.to = function (pos) {
    var that        = this
    var activeIndex = this.getActiveIndex()

    if (pos > (this.$items.length - 1) || pos < 0) return

    if (this.sliding)       return this.$element.one('slid', function () { that.to(pos) })
    if (activeIndex == pos) return this.pause().cycle()

    return this.slide(pos > activeIndex ? 'next' : 'prev', $(this.$items[pos]))
  }

  Carousel.prototype.pause = function (e) {
    e || (this.paused = true)

    if (this.$element.find('.next, .prev').length && $.support.transition.end) {
      this.$element.trigger($.support.transition.end)
      this.cycle(true)
    }

    this.interval = clearInterval(this.interval)

    return this
  }

  Carousel.prototype.next = function () {
    if (this.sliding) return
    return this.slide('next')
  }

  Carousel.prototype.prev = function () {
    if (this.sliding) return
    return this.slide('prev')
  }

  Carousel.prototype.slide = function (type, next) {
    var $active   = this.$element.find('.item.active')
    var $active2   = this.$element.find('.item.active2')
    var $next     = next || $active2[type]()
    var isCycling = this.interval
    var direction = type == 'next' ? 'left' : 'right'
    var fallback  = type == 'next' ? 'first' : 'last'
    var that      = this

    this.sliding = true

    isCycling && this.pause()

    $next = $next.length ? $next : this.$element.find('.item')[fallback]()

    var e = $.Event('slide.bs.carousel', { relatedTarget: $next[0], direction: direction })

    if ($next.hasClass('active')) return

    if ($.support.transition && this.$element.hasClass('slide')) {
      this.$element.trigger(e)
      if (e.isDefaultPrevented()) return
      $next.addClass(type)
      $next[0].offsetWidth // force reflow
      $active.addClass(direction)
      $next.addClass(direction)
      $active
        .one($.support.transition.end, function () {
          $active.removeClass(['active', direction].join(' '))
          $active2.removeClass(['active2', direction].join(' ')).addClass('active')
          $next.removeClass([type, direction].join(' ')).addClass('active2')
          that.sliding = false
          setTimeout(function () { that.$element.trigger('slid') }, 0)
        })
        .emulateTransitionEnd(600)
    } else {
      this.$element.trigger(e)
      if (e.isDefaultPrevented()) return
      $active.removeClass('active')
      $active2.addClass('active')
      $active2.removeClass('active2')
      $next.addClass('active2')
      this.sliding = false
      this.$element.trigger('slid')
    }
    
    var active_index = $('.active').index()
    var active2_index = $('.active2').index()
    
    console.log('.active index: ' + active_index)
    console.log('.active2 index: ' + active2_index)

    if (active2_index < active_index) {
        $('.active').addClass('pull-left');
    } else {
        $('.item').removeClass('pull-left');
    }

    isCycling && this.cycle()

    return this
  }


  // CAROUSEL PLUGIN DEFINITION
  // ==========================

  var old = $.fn.carousel

  $.fn.carousel = function (option) {
    return this.each(function () {
      var $this   = $(this)
      var data    = $this.data('bs.carousel')
      var options = $.extend({}, Carousel.DEFAULTS, $this.data(), typeof option == 'object' && option)
      var action  = typeof option == 'string' ? option : options.slide

      if (!data) $this.data('bs.carousel', (data = new Carousel(this, options)))
      if (typeof option == 'number') data.to(option)
      else if (action) data[action]()
      else if (options.interval) data.pause().cycle()
    })
  }

  $.fn.carousel.Constructor = Carousel


  // CAROUSEL NO CONFLICT
  // ====================

  $.fn.carousel.noConflict = function () {
    $.fn.carousel = old
    return this
  }


  // CAROUSEL DATA-API
  // =================

  $(document).on('click.bs.carousel.data-api', '[data-slide], [data-slide-to]', function (e) {
    var $this   = $(this), href
    var $target = $($this.attr('data-target') || (href = $this.attr('href')) && href.replace(/.*(?=#[^\s]+$)/, '')) //strip for ie7
    var options = $.extend({}, $target.data(), $this.data())
    var slideIndex = $this.attr('data-slide-to')
    if (slideIndex) options.interval = false

    $target.carousel(options)

    if (slideIndex = $this.attr('data-slide-to')) {
      $target.data('bs.carousel').to(slideIndex)
    }

    e.preventDefault()
  })

  $(window).on('load', function () {
    $('[data-ride="carousel"]').each(function () {
      var $carousel = $(this)
      $carousel.carousel($carousel.data())
    })
  })

}(window.jQuery);
